### PR TITLE
Port tan, tanh to datafusion-functions

### PR DESCRIPTION
**Pull request by ongchi**
## Which issue does this PR close?

Part of #9285

### CODE REVIEW

File: datafusion/functions/src/math/tan.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    fn name(&self) -> &str {
+        "tan"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        if arg_types.len() != 1 {
```
Comment (jonahgao) [inline datafusion/functions/src/math/tan.rs:61]: I think this is a duplication with #9401, we can remove it, or use a simple `internal_err`.

File: datafusion/functions/src/math/tanh.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        "tanh"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        if arg_types.len() != 1 {
+            return Err(plan_datafusion_err!(
```
Comment (jonahgao) [inline datafusion/functions/src/math/tanh.rs:62]: The same as above.

File: datafusion/functions/src/math/tan.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+impl TanFunc {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::uniform(
+                1,
+                vec![DataType::Float64, DataType::Float32],
+                Volatility::Immutable,
+            ),
```
Comment (ongchi) [inline datafusion/functions/src/math/tan.rs:42]: > Except for the `generate_signature_error_msg`, everything else looks good to me!

I'm thinking about if the signature has a fixed number of args to 1, is it safe to ignore length checking of `arg_types` in `ScalarUDFImpl::return_type`?

Comment (jonahgao) [reply datafusion/functions/src/math/tan.rs:42]: I think it's safe, that's exactly what was done in [PR 9401](https://github.com/apache/arrow-datafusion/pull/9401/files#diff-2f2aa16b9b3f9d822cac85166bebf32e4f199e0ba5bd7a62f8583181a9487440). We can verify it through a test in sqllogictest.

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @ongchi  and @jonahgao for the review. 🚀

**Pull request merged**